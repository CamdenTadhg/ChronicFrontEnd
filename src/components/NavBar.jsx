import React, {useState, useEffect} from 'react';
import {useNavigate, NavLink} from 'react-router-dom';
import {Navbar, Nav, NavbarBrand, Modal} from 'reactstrap';
import {useSelector} from 'react-redux';
import logo from '../assets/Chronic.png';
import SignupForm from './SignupForm';
import LoginForm from './LoginForm';
import {useDispatch} from 'react-redux';
import {logoutAction} from '../redux/actions/logout';
import ChronicAPI from '../api/chronicAPI';

function NavBar() {
    const [loginOpen, setLoginOpen] = useState(false);
    const [signupOpen, setSignupOpen] = useState(false);
    const navigate = useNavigate();
    const currentUser = useSelector(store => store.profile.userId);
    const dispatch = useDispatch();


    const handleLogout = () => {
        ChronicAPI.token = '';
        dispatch(logoutAction());
        navigate('/');
    }

    const handleHome = () => {
        navigate('/');
    }

    useEffect(() => {
        if (!signupOpen) {
            document.querySelector('.nav-link').blur();
            document.body.focus();
        }
    }, [signupOpen]);

    const toggleLoginModal = () => setLoginOpen(!loginOpen);
    const toggleSignupModal = () => setSignupOpen(!signupOpen);

    return (
        <nav>
            <Navbar color='secondary' full='true' expand='lg' fixed='top'>
                    <NavbarBrand className='brand' onClick={handleHome}>
                        <img className="chronic-logo" src={logo} alt="Logo" />
                        <h6 className='logo-text'>CHRONIC</h6>
                    </NavbarBrand>
                <Nav>
                    {currentUser ?
                        <>
                            <NavLink className="nav-link" to='/profile'>Profile</NavLink>
                            <NavLink className="nav-link" to='/about'>About</NavLink>
                            <button className="nav-link" onClick={handleLogout}>Logout</button>
                        </>:
                        <>
                            <NavLink className="nav-link" to='/about'>About</NavLink>
                            <button className="nav-link" onClick={toggleLoginModal}>Login</button>
                            <button className="nav-link signup" onClick={toggleSignupModal}>Signup</button>
                        </>
                    }
                </Nav>
            </Navbar>
            <Modal isOpen={loginOpen} toggle={toggleLoginModal} size='lg' centered={true}>
                <LoginForm toggleLoginModal={toggleLoginModal}/>
            </Modal>
            <Modal isOpen={signupOpen} toggle={toggleSignupModal} size='lg' centered={true}>
                <SignupForm toggleSignupModal={toggleSignupModal}/>
            </Modal>
        </nav>
    )
}

export default NavBar