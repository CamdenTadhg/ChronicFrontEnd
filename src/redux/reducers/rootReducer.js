import {combineReducers} from 'redux';
import profileReducer from './profileReducer';
import trackingReducer from './trackingReducer';
import dataReducer from './dataReducer';
import latestReducer from './latestReducer';

const rootReducer = combineReducers({
    profile: profileReducer,
    tracking: trackingReducer,
    data: dataReducer,
    latest: latestReducer
});

export default rootReducer;