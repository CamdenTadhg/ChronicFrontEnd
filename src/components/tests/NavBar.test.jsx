import React from 'react';
import {render, fireEvent} from '@testing-library/react';
import {it, expect, vi} from 'vitest';
import '@testing-library/jest-dom';
import {MemoryRouter, useNavigate} from 'react-router-dom';
import rootReducer from '../../redux/reducers/rootReducer';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
import NavBar from '../NavBar';
import ChronicAPI from '../../api/chronicAPI';

const Wrapper = ({children}) => {
    return (
        <MemoryRouter>
            {children}
        </MemoryRouter>
    );
};

vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual('react-router-dom')
    return {
        ...actual,
        useNavigate: vi.fn()
    }
});

it('renders without error', () => {
    const store = configureStore({reducer: rootReducer})
    render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>

            </Wrapper>
        </Provider>);
});

it('matches the snapshot', () => {
    const store = configureStore({reducer: rootReducer})
    const navbar = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>
    );
    expect(navbar).toMatchSnapshot();
});

it('displays the appropriate content to a logged in user', () => {
    const store = configureStore({reducer: rootReducer})
    const {queryByText} = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>);
    expect(queryByText('Login')).toBeInTheDocument();
});

it('displays the appropriate content to a logged out user', () => {
    const store = configureStore({reducer: rootReducer, preloadedState: {profile: {userId: 1}}})
    const {queryByText} = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>);
    expect(queryByText('Logout')).toBeInTheDocument();
});

it('opens the login modal', () => {
    const store = configureStore({reducer: rootReducer})
    const {getByText} = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>);
    fireEvent.click(getByText('Login'));
    expect(getByText('Login Form')).toBeInTheDocument();
});

it('opens the signup modal', () => {
    const store = configureStore({reducer: rootReducer})
    const {getByText} = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>);
    fireEvent.click(getByText('Signup'));
    expect(getByText('Signup Form')).toBeInTheDocument();
});

it('calls the logout function and redirects to home', () => {
    const store = configureStore({reducer: rootReducer, preloadedState:{
        profile: {
            userId: 1,
            email: 'test@test.com',
            name: 'test user',
            isAdmin: false,
            lastLogin: '',
            diagnoses: [],
            symptoms: ['fatigue'],
            medications: ['propanalol'],
            loading: false,
            error: null,
            history: {}
        },
        tracking: {
            primaryTracking: {
                symptoms: {
                    fatigue: 'tracking data here'
                },
                medications: {
                    AM: {
                        propanalol: 'tracking data here'
                    }
                }
            },
            secondaryTracking: {
                symptoms: {
                    fatigue: 'tracking data here'
                },
                medications: {
                    AM: {
                        propanolol: 'tracking data here'
                    }
                }
            },
            loading: false,
            error: null,
            history: {}
        }, 
        data: {
            symptoms: {
                data: 'general data here'
            },
            medications: {
                data: 'general data here'
            },
            loading: false,
            error: null
        },
        latest: {
            articleIds: ['111', '222', '333'],
            loading: false,
            error: null
        }
    } });

    const mockNavigate = vi.fn();
    vi.mocked(useNavigate).mockReturnValue(mockNavigate);
    const {getByText} = render(
        <Provider store={store}>
            <Wrapper>
                <NavBar/>
            </Wrapper>
        </Provider>);
    fireEvent.click(getByText('Logout'));
    expect(getByText('Login')).toBeInTheDocument();
    expect(ChronicAPI.token).toEqual('');
    expect(store.getState()).toEqual({
        profile: {
            userId: null,
            email: '',
            name: '',
            isAdmin: false,
            lastLogin: '',
            diagnoses: [],
            symptoms: [],
            medications: [],
            loading: false,
            error: null,
            history: {}
        },
        tracking: {
            primaryTracking: {
                symptoms: {},
                medications: {}
            },
            secondaryTracking: {
                symptoms: {},
                medications: {}
            },
            loading: false,
            error: null,
            history: {}
        }, 
        data: {
            symptoms: {},
            medications: {},
            loading: false,
            error: null
        },
        latest: {
            articleIds: [],
            loading: true,
            error: null
        }
    });
    expect(mockNavigate).toHaveBeenCalledWith('/');
});

