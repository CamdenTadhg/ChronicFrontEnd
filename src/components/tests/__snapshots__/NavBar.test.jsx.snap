// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav>
        <nav
          class="navbar navbar-expand-lg bg-secondary fixed-top"
          full="true"
        >
          <div
            class="container-fluid"
          >
            <a
              class="brand navbar-brand"
            >
              <img
                alt="Logo"
                class="chronic-logo"
                src="/src/assets/Chronic.png"
              />
              <h6
                class="logo-text"
              >
                CHRONIC
              </h6>
            </a>
            <ul
              class="nav"
            >
              <a
                class="nav-link"
                href="/about"
              >
                About
              </a>
              <button
                class="nav-link"
              >
                Login
              </button>
              <button
                class="nav-link signup"
              >
                Signup
              </button>
            </ul>
          </div>
        </nav>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav>
      <nav
        class="navbar navbar-expand-lg bg-secondary fixed-top"
        full="true"
      >
        <div
          class="container-fluid"
        >
          <a
            class="brand navbar-brand"
          >
            <img
              alt="Logo"
              class="chronic-logo"
              src="/src/assets/Chronic.png"
            />
            <h6
              class="logo-text"
            >
              CHRONIC
            </h6>
          </a>
          <ul
            class="nav"
          >
            <a
              class="nav-link"
              href="/about"
            >
              About
            </a>
            <button
              class="nav-link"
            >
              Login
            </button>
            <button
              class="nav-link signup"
            >
              Signup
            </button>
          </ul>
        </div>
      </nav>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
